### PR TITLE
shortlist feature

### PR DESCRIPTION
**Pull request by elizabeth-jimenez**
Dual Merge: [BE PR](https://github.com/MetaPhase-Consulting/State-TalentMAP-API/pull/505)
~had a git stash scare, so putting here while in-progress~

~blocker: waiting on designs~

Status: Done: Server failing

### CODE REVIEW

File: src/Components/BureauPage/CandidateManagerTable/CandidateManagerTable.jsx [hunk 1]
```diff
@@ -0,0 +1,313 @@
+import PropTypes from 'prop-types';
+import { Link } from 'react-router-dom';
+import InteractiveElement from '../../InteractiveElement';
```
Comment (elizabeth-jimenez) [inline src/Components/BureauPage/CandidateManagerTable/CandidateManagerTable.jsx:3]: This file was for the proposed approach. Hopefully the new approach is preferred and all files having to do with the old approach can be removed.

**Comment by elizabeth-jimenez** [comment]
These tests aren't failing locally(except that one that always fails locally), so not sure what is going on

File: src/Components/BureauPage/BidderRankings/BidderRankings.jsx [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+import { NO_SUBMIT_DATE } from 'Constants/SystemMessages';
+import { fetchBidderRankings } from 'actions/bureauPositionBids';
+import { formatDate, getCustomLocation } from 'utilities';
+import Spinner from '../../Spinner';
+
+const BidderRankings = ({ perdet, cp_id }) => {
+  const bidderRankingData = useSelector(state => state.bureauBidderRankings);
+  const bidderRankingData$ = get(bidderRankingData, perdet) || {};
+  const bidderRankingDataIsLoading = useSelector(state => state.bureauBidderRankingsIsLoading);
+  const bidderRankingDataIsLoading$ = bidderRankingDataIsLoading.has(perdet);
```
Comment (mjoyce91) [inline src/Components/BureauPage/BidderRankings/BidderRankings.jsx:18]: Screen Shot 2021-05-26 at 10 16 56 AM

File: src/Constants/Menu.js [hunk 3]
```diff
@@ -215,6 +205,17 @@ export const GET_PROFILE_MENU = () => MenuConfig([
           'post_user',
         ],
       },
+      checkFlag('flags.static_content') ?
+        {
+          text: 'Candidate Manager',
```
Comment (mjoyce91) [inline src/Constants/Menu.js:210]: Let's increase width of menu

Screen Shot 2021-05-26 at 10 19 43 AM

Comment (elizabeth-jimenez) [reply src/Constants/Menu.js:210]: that candidate manager is likely going to be deleted; going to delete it after they are ok with the shortlist

**Comment by mjoyce91** [comment]
Screen Shot 2021-05-26 at 12 30 36 PM

**Comment by mjoyce91** [comment]
Weird behavior with the spinner. Maybe a position relative vs absolute thing:



https://user-images.githubusercontent.com/11576347/119879341-1f86e380-bef9-11eb-93fa-ed6f8025621c.mov

File: src/Components/BureauPage/BidderRankings/BidderRankings.jsx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+      {
+        showRankingData &&
+          <div className={'bidder-rankings-table-container'}>
+            {
+              bidderRankingDataIsLoading$ &&
+                <Spinner type="bidder-rankings-table" size="small" />
+            }
+            {
+              bidderRankingDataHasErrored$ &&
+              <Alert type="error" title="Error retrieving shortlist bids" />
```
Comment (mjoyce91) [inline src/Components/BureauPage/BidderRankings/BidderRankings.jsx:69]: short list (two words)

File: src/Components/BureauPage/BidderRankings/BidderRankings.jsx [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+                            {item}
+                          </th>
+                        ))
+                      }
+                    </tr>
+                  </thead>
+                  <tbody>
+                    {
+                      !bidderRankingData$.results.length ?
+                        <tr><td>No other bids shortlisted</td></tr>
```
Comment (mjoyce91) [inline src/Components/BureauPage/BidderRankings/BidderRankings.jsx:88]: "Added to short list"?

File: src/Components/BureauPage/BidderRankings/BidderRankings.jsx [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+                            <td>{pos.position.grade}</td>
+                            <td>{pos.bidcycle}</td>
+                            {/* eslint-disable-next-line max-len */}
+                            <td>{pos.submitted_date ? formatDate(pos.submitted_date) : NO_SUBMIT_DATE}</td>
+                          </tr>
+                        ))
+                    }
+                    <tr className="other-sl-count-row">
+                      {/* eslint-disable-next-line react/no-unescaped-entities */}
+                      Number of Bids in other Bureau's ShortLists: {bidderRankingData$['other-sl-bidcount']}
```
Comment (mjoyce91) [inline src/Components/BureauPage/BidderRankings/BidderRankings.jsx:106]: "Short Lists"

Comment (mjoyce91) [inline src/Components/BureauPage/BidderRankings/BidderRankings.jsx:106]: "Number of bids" (i.e. doesn't need to be title cased)

Comment (elizabeth-jimenez) [reply src/Components/BureauPage/BidderRankings/BidderRankings.jsx:106]: ....what about the "ShortLists"? 😏

**Comment by mjoyce91** [comment]
Should we add a link to view the details of this position?

Screen Shot 2021-05-27 at 2 55 12 PM

**Comment by mjoyce91** [comment]
There's a brief flash at 0:11


https://user-images.githubusercontent.com/11576347/119882551-8c4fad00-befc-11eb-8421-8cdcd84cdf3e.mov

**Comment by mjoyce91** [comment]
If we're using this as a deconflict tool, would it make sense to order these bids by rank low (#1) to high?

Screen Shot 2021-05-27 at 3 12 12 PM

**Comment by mjoyce91** [comment]
Can we hook into the `onDrag` event and hide this section when it gets dragged?


https://user-images.githubusercontent.com/11576347/119884330-822eae00-befe-11eb-92be-a2b943fc925f.mov

File: src/Components/BureauPage/PositionManagerBidders/PositionManagerBidders.jsx [hunk 7]
```diff
@@ -343,31 +345,42 @@ class PositionManagerBidders extends Component {
...
+
+    const shortListSection = (
+      <>
+        <div className="list-toggle-container">
+          <InteractiveElement title="Toggle visibility" onClick={() => this.toggleVisibility('shortListVisible')}><FA name={shortListVisible ? 'chevron-down' : 'chevron-up'} /></InteractiveElement>
+          <h3>Short List ({this.state.shortList.length})</h3>
+          {shortListLock}
+        </div>
+        {
+          shortListVisible &&
```
Comment (elizabeth-jimenez) [inline src/Components/BureauPage/PositionManagerBidders/PositionManagerBidders.jsx:383]: eslint was complaining. Only changed indents here

File: src/Components/BureauPage/PositionManagerBidders/PositionManagerBidders.jsx [hunk 8]
```diff
@@ -488,7 +511,13 @@ class PositionManagerBidders extends Component {
                                       isDragDisabled={dndDisabled}
                                     >
                                       {(provided$, snapshot$) => (
+                                        // eslint-disable-next-line max-len
+                                        // eslint-disable-next-line jsx-a11y/interactive-supports-focus
                                         <div
+                                          role="button"
+                                          /* eslint-disable-next-line max-len */
+                                          onMouseDown={(e) => { this.handleEvent(e, item.bid.emp_id); }}
+                                          onMouseUp={this.handleEvent}
```
Comment (elizabeth-jimenez) [inline src/Components/BureauPage/PositionManagerBidders/PositionManagerBidders.jsx:520]: 🤮 🤷‍♀️

**Comment by elizabeth-jimenez** [comment]
> There's a brief flash at 0:11
> 
>  Screen.Recording.2021-05-27.at.3.01.20.PM.mov

Done 😁

**Comment by elizabeth-jimenez** [comment]
> Can we hook into the `onDrag` event and hide this section when it gets dragged?
> 
>  Screen.Recording.2021-05-27.at.3.15.13.PM.mov

Not working perfectly, but i think it's better than before

File: src/Components/BureauPage/BidderRankings/BidderRankings.jsx [hunk 9]
```diff
@@ -51,7 +51,7 @@ const BidderRankings = ({ perdet, cp_id, is_dragging, is_mouse_down, mouse_down_
 
   useEffect(() => {
     if (isEqual(mouse_down_emp, perdet)) {
-      if (is_dragging || is_mouse_down) {
```
Comment (elizabeth-jimenez) [inline src/Components/BureauPage/BidderRankings/BidderRankings.jsx]: changing this to `||` gives us the desired effect, but then we can't click on the position details link

File: src/Components/BureauPage/PositionManagerBidders/PositionManagerBidders.jsx [hunk 10]
```diff
@@ -69,6 +69,8 @@ const getClassificationsTooltip = (classifications) => (
 
 const getItemStyle = (isDragging, draggableStyle) => {
   const border = isDragging ? '1px solid black' : 'none';
+  const height = isDragging ? '130px' : '';
+  const overflowY = isDragging ? 'hidden' : '';
```
Comment (elizabeth-jimenez) [inline src/Components/BureauPage/PositionManagerBidders/PositionManagerBidders.jsx:73]: unfortunately, I don't think this does anything(might be applied too slowly)

File: src/actions/bureauPositionBids.js [hunk 11]
```diff
@@ -238,3 +238,9 @@ export function fetchBidderRankings(perdet, cp_id) {
       });
   };
 }
+
+export function clearBidderRankings() {
+  return (dispatch) => {
+    dispatch(bidderRankingsFetchSuccess(null, null, true));
+  };
+}
```
Comment (elizabeth-jimenez) [inline src/actions/bureauPositionBids.js:246]: this is to clear the blink you saw, Mike

File: src/Components/BureauPage/PositionManagerBidders/PositionManagerBidders.jsx [hunk 12]
```diff
@@ -342,31 +349,42 @@ class PositionManagerBidders extends Component {
...
-            <InteractiveElement title="Toggle visibility" onClick={() => this.toggleVisibility('shortListVisible')}><FA name={shortListVisible ? 'chevron-down' : 'chevron-up'} /></InteractiveElement>
-            <h3>Short List ({this.state.shortList.length})</h3>
-            {shortListLock}
-          </div>
-          {
-            shortListVisible &&
+  handleEvent = (event, id) => {
+    if (event.type === 'mousedown') {
+      this.setState({ isMouseDown: true });
+      this.setState({ mouseDownEmp: id });
```
Comment (mjoyce91) [inline src/Components/BureauPage/PositionManagerBidders/PositionManagerBidders.jsx:355]: setStates can be combined into one (avoids an extra render):

```jsx
this.setState({ isMouseDown: true, mouseDownEmp: id });
```

File: src/Components/BureauPage/PositionManagerBidders/PositionManagerBidders.jsx [hunk 13]
```diff
@@ -342,31 +349,42 @@ class PositionManagerBidders extends Component {
...
+    }
+    if (event.type === 'mouseup') {
+      this.setState({ isMouseDown: false });
+      this.setState({ mouseDownEmp: '' });
```
Comment (mjoyce91) [inline src/Components/BureauPage/PositionManagerBidders/PositionManagerBidders.jsx:359]: setStates can be combined into one (avoids an extra render):

```jsx
this.setState({ isMouseDown: false, mouseDownEmp: '' });
```

Comment (elizabeth-jimenez) [reply src/Components/BureauPage/PositionManagerBidders/PositionManagerBidders.jsx:359]: oohh very nice 🥇

**Review by mjoyce91 — APPROVED** [review]

**Pull request merged**